python/perf-kernels/flash-attention.py

```diff
@@ -127,6 +164,19 @@ def _attn_fwd_inner(
         m_ij = tl.maximum(m_i, tl.max(qk, 1))
         qk = qk - m_ij[:, None]
         p = tl.math.exp2(qk)
+        # CAVEAT: Must update l_ij before applying dropout
+        l_ij = tl.sum(p, 1)
+        # Note about the conflicts of Flash attention algorithm and PyTorch's CUDA implementation
+        # PyTorch needs to return softmax(qk) (dropout mask encoded in sign bits)
+        # While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)
+        if ENABLE_DROPOUT:
+            philox_offset = batch_philox_offset + start_m * BLOCK_M * seqlen_k + start_n
+            keep = dropout_mask(philox_seed, philox_offset, dropout_p, BLOCK_M, BLOCK_N, seqlen_k)
+            if RETURN_ENCODED_SOFTMAX:
+                tl.store(encoded_softmax_block_ptr, tl.where(keep, p, -p).to(encoded_softmax_block_ptr.type.element_ty))
+            p = tl.where(keep, p, 0.0)
+        elif RETURN_ENCODED_SOFTMAX:
```

Comment:
Return_softmax implies dropout but converse is not true. Can we instead have

```
if return_softmax:
  apply dropout
  store softmax
else if dropout:
  apply_dropout
```

Reply:
Nope, `return_softmax` is orthogonal to `dropout`. User can request softmax results even if dropout is disabled.
For example, we probably want to take a look at softmax values when debugging https://github.com/pytorch/pytorch/issues/116176

Reply:
This is not how CUDA does it:

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L419

This also brings me back to my comment below on adding a comment as to why this feature is needed. IT is not clear.

Reply:
It is necessary for SDPA testing and integration into PyTorch.  The behavior we have is correct as per unit tests but it does diverge from this example.  From our perspective we are making sure we have correct behavior but do not need to 100% match the CUTLASS reference.

Reply:
> This is not how CUDA does it:
> 
> https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L419
> 
> This also brings me back to my comment below on adding a comment as to why this feature is needed. IT is not clear.

For `encoded_softmax` we are not going to follow CUDA's behavior, and if I read that correctly the two choices are still orthogonal. `apply_dropout` will be a noop if `dropout_p` == 0 @groenenboomj

Reply:
If pytorch is okay not following CUDA, that works for me.

Also, no, I don't think they're not orthogonal according to the implementation:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L604